feat(tracing): add tracing spans to all I/O sections

fvaleye opened the pull request:
# Description
To better understand performance in the `delta-rs` crate, I added additional tracing to capture more detailed debug-level performance information.

Python now uses `OpenTelemetry` to collect tracing data emitted from Rust.
With this change, we gain true end-to-end visibility: Python spans can serve as parents of Rust spans (and vice versa), ensuring a continuous trace across both runtimes.

# Related Issue(s)
-  close #3641

# Documentation
- https://docs.rs/tracing/latest/tracing/

rtyler commented on crates/core/src/kernel/transaction/mod.rs:
```diff
@@ -619,14 +619,21 @@ impl<'a> std::future::IntoFuture for PreparedCommit<'a> {
             let commit_or_bytes = this.commit_or_bytes;
 
             if this.table_data.is_none() {
+                tracing::debug!("committing initial table version 0");
```
:wave: in a lot of places we actually just import `use tracing::*;` which means this module prefix isn't necessary.

IMHO it's reasonable for us to pull `tracing::*` into most modules :smile:

fvaleye replied:
Just imported with `use tracing::*;`, if you found it's ok for readability.
I didn't want to mix `log` and `tracing` crates, but it seems reasonable for now! We can tighten the imports or add explicit bridging later if the project grows.

rtyler commented on crates/core/src/delta_datafusion/mod.rs:
```diff
@@ -2388,4 +2442,86 @@ mod tests {
+        span.record("matching_files", 12);
+    }
+
+    #[test]
+    fn test_find_files_tracing_log_events() {
+        tracing::debug!("using partition-only scan (memory table)");
+        tracing::debug!("using physical data scan");
+        tracing::debug!("no predicate, returning all files");
+
+        tracing::info!(
+            candidates = 45,
+            scan_type = "partition",
+            "file pruning completed"
+        );
+
+        tracing::info!(
+            candidates = 12,
+            scan_type = "physical",
+            "file pruning completed"
+        );
+
+        tracing::debug!(matching_files = 12, "physical scan completed");
+    }
```
:laughing: this test doesn't actually do anything

fvaleye replied:
I'm working on it... I removed some examples where I tested the execution.
I wanted to see the test coverage results first.

rtyler commented on crates/core/src/kernel/transaction/mod.rs:
```diff
@@ -967,4 +1026,58 @@ mod tests {
+    }
+
+    #[test]
+    fn test_commit_tracing_log_events() {
+        tracing::debug!("committing initial table version 0");
+        tracing::info!(
+            version = 6,
+            num_retries = 1,
+            "transaction committed successfully"
+        );
+        tracing::warn!(
+            base_version = 5,
+            latest_version = 7,
+            "table updated during transaction"
+        );
+        tracing::error!(version = 8, "commit failed, aborting");
+        tracing::debug!(conflicts_checked = 3, "all conflicts resolved");
+    }
```
This test also doesn't seem to actually do anything :upside_down_face:

fvaleye replied:
I'm working on it... I removed some examples where I tested the execution.
I wanted to see the test coverage results first.

rtyler reviewed (dismissed):
Generally very excited to have more tracing in place. Once tests pass I'm comfortable merging this

ion-elgreco commented:
This is stdout/err tracing only, right?

fvaleye commented:
> This is stdout/err tracing only, right?

I used `warning`, `info`, and `debug` levels together with `tracing`, adding `Spans` with custom fields to track performance.

`tracing` isn’t limited to stdout or stderr. It’s a subscriber-agnostic instrumentation framework that emits structured events with typed fields, not just plain text logs. 

Where these traces go is entirely up to the user: with `tracing-subscriber`, traces can be sent to distributed tracing backends like Jaeger, Zipkin, Tempo, Honeycomb, or Datadog, to OpenTelemetry collectors, to custom backends, or nowhere at all.

Example:
```
// Option 1: Console output
let subscriber = tracing_subscriber::fmt()
    .with_env_filter("deltalake=debug")
    .finish();
tracing::subscriber::set_global_default(subscriber)?;

```

Side note: Importantly, if no subscriber is configured, the overhead is nearly zero.

rtyler commented:
I'm marking this as a draft, please change that when you're ready for this to be merged @fvaleye

fvaleye commented:
Ready for review @rtyler @ion-elgreco 

@ion-elgreco I added what was required from Python to collect data from `tracing` crate using `OpenTelemetry`.

**Side notes:** I fixed `typos` issues, but the CI is still failling...

ion-elgreco commented on python/src/lib.rs:
```diff
@@ -2159,6 +2162,39 @@ fn rust_core_version() -> &'static str {
+            })
+        })
+    }
+    #[cfg(not(feature = "otel"))]
```
We don't really not feature gates for python tbh

fvaleye commented on python/src/lib.rs:
```diff
@@ -2159,6 +2162,39 @@ fn rust_core_version() -> &'static str {
+fn init_tracing(endpoint: String) -> PyResult<()> {
+    #[cfg(feature = "otel")]
+    {
+        let rt = Runtime::new().map_err(|e| {
```
Yes, thanks!

abhiaagarwal commented on crates/core/src/tracing_otlp.rs:
Why does this live in the `deltalake_core` crate inside of the python implementation? There are more idiomatic ways to do this in rust-land, and if you've already initialized tracing in your app for other components, this will conflict

fvaleye replied:
It's a valid point @abhiaagarwal, thanks!
There is no need to include this part in the `deltalake_core`, as it's more related to the Python binding. I will move this part.

abhiaagarwal commented on crates/core/src/kernel/transaction/mod.rs:
```diff
@@ -637,9 +644,20 @@ impl<'a> std::future::IntoFuture for PreparedCommit<'a> {
+                target_version = field::Empty,
+                conflicts_checked = 0
+            );
+            let _enter = commit_span.enter();
```
thanks for the quick responses :) I've been instrumenting delta in production through hacky means, very excited to get more metrics!

ion-elgreco commented:
@fvaleye is this in a state it can be reviewed again?

fvaleye commented:
> @fvaleye is this in a state it can be reviewed again?

Yes!

fvaleye commented on crates/core/src/kernel/transaction/mod.rs:
```diff
@@ -619,14 +619,21 @@ impl<'a> std::future::IntoFuture for PreparedCommit<'a> {
                     version: 0,
                     data: this.data,
-                    create_checkpoint: false,
-                    cleanup_expired_logs: None,
+                    create_checkpoint: this
+                        .post_commit
+                        .map(|v| v.create_checkpoint)
+                        .unwrap_or_default(),
+                    cleanup_expired_logs: this
+                        .post_commit
+                        .map(|v| v.cleanup_expired_logs)
+                        .unwrap_or_default(),
```
Oh, it was part of another [investigation](https://github.com/delta-io/delta-rs/issues/3780)!
Thanks, removing it 😄
